### PR TITLE
Introspection configuration per replica

### LINKED ISSUE #12660: compute,adapter: push introspection configuration down to replica
In the switch to explicit cluster replica management, we didn't think through what to do with the options for the introspection sources. Currently introspection is configured per cluster:

```sql
CREATE CLUSTER c INTROSPECTION GRANULARITY '1s'
```

We have the ability to control introspection per _replica_ though, rather than per cluster. We probably want to expose that feature, since it might be nice to leave introspection disabled unless you're actively doing performance debugging, at which point you could turn on a new replica with introspection enabled.

~Does that seem right to folks?~ ~The specific proposal is to allow _overriding_ the introspection options per replica. They can still be configured at the cluster level, which sets the default for any replicas that do not otherwise have their introspection options configured.~

Abridged proposal: we should only allow configuring the introspection options per replica, not per cluster. We can later figure out if it makes sense to allow specifying them per cluster, and propagate that default down to replicas that didn't explicitly override them. But we're short on time. 

cc @sploiselle @jkosh44 @teskje @antiguru

### PR DESCRIPTION
**Pull request by teskje**
This PR moves the configuration for compute introspection from the cluster-level to the replica-level. In the future we may want to include some kind of configuration inheritance, with default settings for the cluster that can get overwritten per replica. We are punting on this now because of how close we are to EA. Adding back cluster-level introspection configuration can be done later without breaking backwards-compatibility.

The one thing that is a bit messy now is that the old arranged introspection sources share their catalog items (indexes) between all replicas of a cluster. We need to create those when we create the cluster, even though not all replicas may have introspection enabled, and therefore may not maintain arrangements backing the catalog indexes. So the adapter needs to check before serving a peek that targets arranged introspection sources, whether the targeted replica actually has introspection enabled. Once we manage to make the persisted introspection sources ergonomic, we plan to remove the arranged ones, so this ugliness will go away eventually.

Also note that this PR changes the format of the cluster and replica configs we store in stash. This would need a migration, but writing one seems hard. So I'd rather wait until next Wednesday and then merge this PR in time for the next release that allows breaking changes.

### Motivation

  * This PR adds a known-desirable feature.

Closes #12660.

### Tips for reviewer

I tried to split this into several more digestible commits, but I'm not sure if I succeeded. I'd suggest you try looking at the commits separately and see if that helps.

### Checklist

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.
- [ ] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-protobuf` label.
- [ ] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - Move configuration of introspection from the cluster-level to the replica-level.

### CODE REVIEW

**Review by mjibson — APPROVED** [review]
sql/catalog/coord/stash parts lgtm

**Comment by benesch** [comment]
> Also note that this PR changes the format of the cluster and replica configs we store in stash. This would need a migration, but writing one seems hard. So I'd rather wait until next Wednesday and then merge this PR in time for the next release that allows breaking changes.

SGTM. I've added it to the release playbook so we don't forget: https://www.notion.so/materialize/Playbook-3827577236a44cc39b885116554cf200#c1ecae0669174fc9b97790e03e943c0d

File: src/adapter/src/catalog.rs [hunk 1]
```diff
@@ -3533,19 +3501,14 @@ impl<S: Append> Catalog<S> {
...
                     name,
-                    config,
                     arranged_introspection_sources,
                 } => {
                     if is_reserved_name(&name) {
                         return Err(AdapterError::Catalog(Error::new(
                             ErrorKind::ReservedClusterName(name),
                         )));
                     }
-                    let id = tx.insert_compute_instance(
```
[the remark this answers is not in the document]
Comment (teskje) [reply src/adapter/src/catalog.rs:3544]: Thanks for the heads-up!

**Review by lluki — APPROVED** [review]
LGTM

File: test/sqllogictest/select_introspection.slt [hunk 2]
```diff
@@ -81,18 +80,30 @@ DROP CLUSTER REPLICA test.replica_b;
...
 statement ok
 DROP CLUSTER test CASCADE
 
 statement ok
 CREATE CLUSTER test
-  REPLICAS (replica_a (SIZE '1')),
-  INTROSPECTION INTERVAL 0
+  REPLICAS (
+    replica_a (SIZE '1', INTROSPECTION INTERVAL 0),
+    replica_b (SIZE '1', INTROSPECTION INTERVAL 1)
```
Comment (philip-stoev) [inline test/sqllogictest/select_introspection.slt:94]: Is a naked number without a specifier interpreted as seconds or something else? Maybe all values except 0 should be disallowed.

**Review by philip-stoev — APPROVED** [review]
Thank you. Unfortunately I am buried right now so can not test this myself. Test changes and additions LGTM.

File: test/sqllogictest/select_introspection.slt (hunk 2 above)
Comment (teskje) [reply test/sqllogictest/select_introspection.slt:94]: Numbers are interpreted by casting them to a string and `INTERVAL '1'` means one second.

```
materialize=> SELECT INTERVAL '1';
 interval
----------
 00:00:01
(1 row)
```

I agree that this is not obvious, but it is consistent with Postgres. We could think about not doing the implicit cast from number to string though.

**Comment by teskje** [comment]
Well, I guess I was lucky :)

**Pull request merged**